mempool/src/pool/mod.rs

```diff
@@ -968,9 +970,36 @@ impl<M: MemoryUsageEstimator> Mempool<M> {
             .expect("best index to exist");
         let tx_source = TransactionSourceForConnect::for_mempool(&best_index);
 
+        let mut unique_txids = BTreeSet::new();
+        let mut ordered_txids = vec![];
+
+        for transaction in tx_accumulator.transactions().iter() {
+            // Use transactions already in the Accumulator to sort for
+            // uniqueness only i.e don't send them through the
+            // verifier
+            unique_txids.insert(transaction.transaction().get_id());
+        }
+
+        for transaction_id in transaction_ids {
+            if unique_txids.insert(transaction_id) {
+                ordered_txids.push(transaction_id)
+            }
+        }
+
+        if fill_from_mempool {
+            let fill_txids = self.store.txs_by_ancestor_score.iter().map(|(_, id)| id).rev();
+
+            for transaction_id in fill_txids {
+                if unique_txids.insert(*transaction_id) {
+                    ordered_txids.push(*transaction_id)
+                }
+            }
```

Comment:
This is not good. The original algorithm carefully used iterators specifically to avoid iterating over all transactions in the mempool, of which there can be many. That applies especially to sophisticated node operators that may want to increase the mempool size limit so they lose fewer transactions, e.g. stakers. However here, the whole score map is iterated over first to fill a vec but only a tiny fraction of it is used if mempool is anywhere close to full.

I would say in general, try to avoid filling in temporary data structures if using an iterator directly will do just fine.

Reply:
Ah, the `fuse()` below. Nice.

Reply:
I don't think this has been addressed (as of 1351109e697df9f). More things are implemented in terms of iterators. However ultimately, it still has to iterate the whole collection in order to extend the `ordered_txids` vector. Ideally this is avoided so the loop below takes elements from the iterator directly. That way, we will not visit all tx ids in mempool when an error or full accumulator interrupts the iteration.